File: simulation/core/contracts/domain.py
"""Domain contracts — My Digital Twin Garden (TASK 002). No physics implemented."""
from __future__ import annotations
from datetime import datetime
from typing import Optional, Literal, List
from pydantic import BaseModel, Field

# --- Spatial / Reference ---
class GardenReference(BaseModel):
    """G0 local origin and orientation."""
    origin: list[float] = Field(default=[0.0, 0.0, 0.0], description="G0 position")
    north_bearing_deg: float = Field(default=0.0)
    coord_system: Literal["local-garden"] = "local-garden"

class SpatialTransform(BaseModel):
    position: list[float] = [0.0, 0.0, 0.0]
    rotation_euler_deg: list[float] = [0.0, 0.0, 0.0]
    scale: list[float] = [1.0, 1.0, 1.0]
    parent_id: Optional[str] = None

# --- Garden / Boundary ---
class Boundary(BaseModel):
    points: list[list[float]]  # ordered polygon (L-shape allowed)
    type: Literal["polygon", "other"] = "polygon"

class Garden(BaseModel):
    id: str
    reference: GardenReference
    boundary: Boundary
    objects: List[str] = Field(default_factory=list)

# --- Spatial objects ---
class SpatialObject(BaseModel):
    id: str
    type: Literal["building","fence","tree","rack","tier","container","container_cell","plant","other"]
    transform: SpatialTransform
    parent_id: Optional[str] = None

class Building(SpatialObject):
    pass

class FenceWall(SpatialObject):
    pass

class TreeObstacle(SpatialObject):
    pass

# --- Rack / Tier / Container ---
class RackTier(BaseModel):
    id: str
    parent_rack_id: str
    index: int
    transform: SpatialTransform

class Container(BaseModel):
    id: str
    container_type: Literal["polybag","pot","seedling_tray","ground","other"]
    dimensions: Optional[list[float]] = None  # [w,h,d]
    volume_l: Optional[float] = None
    material: Optional[str] = None
    color: Optional[str] = None
    growing_medium_ref: Optional[str] = None
    drainage_ref: Optional[str] = None
    transform: SpatialTransform
    parent_id: Optional[str] = None
    cell_ids: List[str] = Field(default_factory=list)

class ContainerCell(BaseModel):
    id: str
    parent_container_id: str
    index: int
    transform: SpatialTransform
    occupied_plant_id: Optional[str] = None

# --- Plant (identity) ---
class Plant(BaseModel):
    id: str
    species_id: Optional[str] = None
    variety_id: Optional[str] = None
    current_container_id: Optional[str] = None
    lifecycle_stage: Optional[str] = None  # planned; not computed

# --- PlantState (separate from identity/architecture) ---
class PlantState(BaseModel):
    plant_id: str
    timestamp: datetime
    age_days: Optional[float] = None  # planned; unknown if missing
    phenological_stage: Optional[str] = None
    biomass_g: Optional[float] = None
    carbon_pool_g: Optional[float] = None
    water_status: Optional[str] = None  # planned
    nutrient_status: Optional[str] = None
    stress_state: Optional[str] = None
    # unknown / not-applicable / zero must be explicit (not silent)

# --- PlantArchitecture (no Three.js) ---
class PlantOrgan(BaseModel):
    id: str
    organ_type: Literal["axis","internode","leaf","bud","flower","fruit","root","other"]
    topology_ref: Optional[str] = None

class PlantArchitecture(BaseModel):
    plant_id: str
    topology: Optional[str] = None
    organs: List[PlantOrgan] = Field(default_factory=list)
    geometry_metadata: Optional[dict] = None

# --- Variety ---
class VarietyParameter(BaseModel):
    name: str
    value: Optional[float] = None  # None = unknown / missing
    unit: Optional[str] = None  # required when value known; None if unknown
    source: Literal["literature","measured","fitted_inferred","assumption_placeholder","unknown"] = "unknown"
    method: Optional[str] = None
    uncertainty: Optional[float] = None
    valid_range: Optional[list] = None  # [min, max]
    version: Optional[str] = None
    provenance: Optional[str] = None

class VarietyProfile(BaseModel):
    species: str
    cultivar: Optional[str] = None
    description: Optional[str] = None
    parameter_collection: List[VarietyParameter] = Field(default_factory=list)

# --- Environment / Light ---
class EnvironmentState(BaseModel):
    timestamp: datetime
    solar_state: Optional[str] = None  # planned; not computed
    temperature_c: Optional[float] = None  # planned
    humidity_pct: Optional[float] = None
    precipitation_mm: Optional[float] = None
    water_availability: Optional[str] = None  # planned
    nutrient_environment: Optional[str] = None

class LightField(BaseModel):
    timestamp: datetime
    direct_component: Optional[float] = None  # planned
    diffuse_component: Optional[float] = None
    reflected_component: Optional[float] = None
    illuminance_lux: Optional[float] = None  # observed; not = photosynthesis
    estimated_plant_relevant: Optional[float] = None  # planned
    spatial_association: Optional[str] = None  # e.g., organ/plant ref

# --- Measurement / Observation ---
class Measurement(BaseModel):
    id: str
    timestamp: datetime
    variable: str  # extensible vocabulary; e.g. illuminance, temperature, soil_moisture
    value: float
    unit: str  # required; never silent; lux stays lux; never converted to PPFD
    location_target_id: Optional[str] = None  # named SpatialObject / container / plant
    spatial_ref: Optional[dict] = None  # e.g. {"x":0,"y":0,"z":0} or garden ref; not hard-coded dimensions
    instrument: Optional[str] = None  # instrument name/type; do not invent serial numbers
    instrument_id: Optional[str] = None  # identifier only if actually known
    method: Optional[str] = None
    calibration_ref: Optional[str] = None  # only if actually known; never fabricated
    observer_source: Optional[str] = None  # observer or data source
    uncertainty: Optional[float] = None  # known = value; unknown/absent = None, NOT zero
    provenance: Optional[str] = None  # source, method, version/date, notes
    quality_flag: Optional[str] = None  # optional; not statistical QC yet
    notes: Optional[str] = None  # raw-context preservation; do not smooth/remove
    is_synthetic_example: bool = False  # synthetic fixtures must be explicitly labeled

class Observation(BaseModel):
    id: str
    timestamp: datetime
    target_id: Optional[str] = None  # plant / container / object reference
    observation_type: Optional[str] = None  # e.g. phenological_stage, status, event
    category: Optional[str] = None  # qualitative / photographic / event / structured
    content: str  # human-readable; non-numeric observations preserved here
    structured_attributes: Optional[dict] = None  # structured payload; numeric only when appropriate and with unit
    value_numeric: Optional[float] = None  # only when observation includes a measurement-like number; never assume
    unit: Optional[str] = None  # required when value_numeric present; lux stays lux
    location_target_id: Optional[str] = None
    spatial_ref: Optional[dict] = None
    method: Optional[str] = None
    observer_source: Optional[str] = None
    uncertainty: Optional[float] = None  # only when applicable; unknown = None
    notes: Optional[str] = None  # provenance / context; do not fabricate
    provenance: Optional[str] = None
    quality_flag: Optional[str] = None
    is_synthetic_example: bool = False

# --- Events ---
class InterventionEvent(BaseModel):
    id: str
    timestamp: datetime
    event_type: Literal["watering","fertilization","transplanting","pruning","repotting","harvesting","other"]
    target_id: str
    parameters: Optional[dict] = None
    user_source: Optional[str] = None

# --- Scenario / Simulation ---
class Scenario(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    base_snapshot_ref: Optional[str] = None
    time_range: Optional[list] = None  # [start, end] planned
    assumptions: List[str] = Field(default_factory=list)
    interventions: List[str] = Field(default_factory=list)  # event refs
    parameter_overrides: Optional[dict] = None
    random_seed: Optional[int] = None

class SimulationState(BaseModel):
    simulation_time: float  # planned; float seconds/days
    environment_state: Optional[EnvironmentState] = None
    soil_water_state_ref: Optional[str] = None
    plant_states: List[str] = Field(default_factory=list)
    architectures_ref: List[str] = Field(default_factory=list)
    pending_events: List[str] = Field(default_factory=list)

class SimulationRun(BaseModel):
    id: str
    source_state_ref: str
    scenario_ref: Optional[str] = None
    model_version_ref: Optional[str] = None
    parameter_set_version_ref: Optional[str] = None
    random_seed: Optional[int] = None
    forecast_horizon: Optional[float] = None
    assumptions: List[str] = Field(default_factory=list)
    checkpoints_ref: Optional[List[str]] = None
    uncertainty_refs: Optional[List[str]] = None

class Prediction(BaseModel):
    id: str
    source_snapshot_ref: str
    forecast_time: float
    plant_states_ref: List[str] = Field(default_factory=list)
    architectures_ref: List[str] = Field(default_factory=list)
    scenario_ref: Optional[str] = None
    # Not mutable live Garden state; separate result representation

# --- Model registry contracts ---
class ModelVersion(BaseModel):
    model_name: str
    version: str  # semantic
    component_versions: Optional[dict] = None
    equations_ref: Optional[str] = None
    dependency_metadata: Optional[dict] = None

class ParameterSetVersion(BaseModel):
    parameter_set_id: str
    version: str
    source_ref: Optional[str] = None
    applicability_scope: Optional[str] = None

File: frontend/src/contracts/domain.ts
/* Domain contracts — frontend consumption. No Three.js, no physics. */

export interface GardenReference {
  origin: [number, number, number];
  north_bearing_deg: number;
  coord_system: "local-garden";
}

export interface SpatialTransform {
  position: [number, number, number];
  rotation_euler_deg: [number, number, number];
  scale: [number, number, number];
  parent_id?: string;
}

export interface Boundary {
  points: [number, number, number][]; // ordered polygon
  type: "polygon" | "other";
}

export interface Garden {
  id: string;
  reference: GardenReference;
  boundary: Boundary;
  objects: string[];
}

export interface SpatialObject {
  id: string;
  type: "building" | "fence" | "tree" | "rack" | "tier" | "container" | "container_cell" | "plant" | "other";
  transform: SpatialTransform;
  parent_id?: string;
}

export interface Container {
  id: string;
  container_type: "polybag" | "pot" | "seedling_tray" | "ground" | "other";
  dimensions?: [number, number, number];
  volume_l?: number;
  material?: string;
  color?: string;
  growing_medium_ref?: string;
  drainage_ref?: string;
  transform: SpatialTransform;
  parent_id?: string;
  cell_ids: string[];
}

export interface ContainerCell {
  id: string;
  parent_container_id: string;
  index: number;
  transform: SpatialTransform;
  occupied_plant_id?: string;
}

export interface Plant {
  id: string;
  species_id?: string;
  variety_id?: string;
  current_container_id?: string;
  lifecycle_stage?: string;
}

export interface PlantState {
  plant_id: string;
  timestamp: string; // ISO-8601
  age_days?: number; // unknown if missing; do not default to 0
  phenological_stage?: string;
  biomass_g?: number;
  carbon_pool_g?: number;
  water_status?: string;
  nutrient_status?: string;
  stress_state?: string;
}

export interface PlantOrgan {
  id: string;
  organ_type: "axis" | "internode" | "leaf" | "bud" | "flower" | "fruit" | "root" | "other";
  topology_ref?: string;
}

export interface PlantArchitecture {
  plant_id: string;
  topology?: string;
  organs: PlantOrgan[];
  geometry_metadata?: Record<string, unknown>;
}

export interface VarietyParameter {
  name: string;
  value?: number; // unknown if missing; not silently zero
  unit?: string; // required when value known
  source: "literature" | "measured" | "fitted_inferred" | "assumption_placeholder" | "unknown";
  method?: string;
  uncertainty?: number;
  valid_range?: [number, number];
  version?: string;
  provenance?: string;
}

export interface VarietyProfile {
  species: string;
  cultivar?: string;
  description?: string;
  parameter_collection: VarietyParameter[];
}

export interface EnvironmentState {
  timestamp: string;
  solar_state?: string; // planned
  temperature_c?: number;
  humidity_pct?: number;
  precipitation_mm?: number;
  water_availability?: string;
  nutrient_environment?: string;
}

export interface LightField {
  timestamp: string;
  direct_component?: number;
  diffuse_component?: number;
  reflected_component?: number;
  illuminance_lux?: number; // observed; NOT photosynthesis
  estimated_plant_relevant?: number;
  spatial_association?: string;
}

export interface Measurement {
  id: string;
  timestamp: string;
  variable: string;
  value: number;
  unit: string; // explicit; never silent
  location_target_id?: string;
  instrument?: string;
  method?: string;
  uncertainty?: number;
  provenance?: string;
}

export interface Observation {
  id: string;
  timestamp: string;
  target_id?: string;
  category?: string;
  content: string;
  structured_attributes?: Record<string, unknown>;
  observer_source?: string;
}

export interface InterventionEvent {
  id: string;
  timestamp: string;
  event_type: "watering" | "fertilization" | "transplanting" | "pruning" | "repotting" | "harvesting" | "other";
  target_id: string;
  parameters?: Record<string, unknown>;
  user_source?: string;
}

export interface Scenario {
  id: string;
  name: string;
  description?: string;
  base_snapshot_ref?: string;
  time_range?: [string, string]; // planned
  assumptions: string[];
  interventions: string[]; // event refs
  parameter_overrides?: Record<string, unknown>;
  random_seed?: number;
}

export interface SimulationState {
  simulation_time: number;
  environment_state?: EnvironmentState;
  soil_water_state_ref?: string;
  plant_states: string[];
  architectures_ref: string[];
  pending_events: string[];
}

export interface SimulationRun {
  id: string;
  source_state_ref: string;
  scenario_ref?: string;
  model_version_ref?: string;
  parameter_set_version_ref?: string;
  random_seed?: number;
  forecast_horizon?: number;
  assumptions: string[];
  checkpoints_ref?: string[];
  uncertainty_refs?: string[];
}

export interface Prediction {
  id: string;
  source_snapshot_ref: string;
  forecast_time: number;
  plant_states_ref: string[];
  architectures_ref: string[];
  scenario_ref?: string;
}

export interface ModelVersion {
  model_name: string;
  version: string;
  component_versions?: Record<string, string>;
  equations_ref?: string;
  dependency_metadata?: Record<string, unknown>;
}

export interface ParameterSetVersion {
  parameter_set_id: string;
  version: string;
  source_ref?: string;
  applicability_scope?: string;
}
